proposals/accepted/051-ghc-base-libraries.rst

```diff
@@ -0,0 +1,330 @@
+   This is quite open-ended; it is not clear what would be desirable, or how much it would cost.
+
+Some modules in ``ghc-internals`` will very directly affect exports of ``base`` (e.g via shim).
+These modules could be identified, via the existing ``CODEOWNERS`` mechanism, to ping CLC on any commit to those modules.
+This list could be selective, or include all of ``ghc-internals``, at CLC's preference.
+
+Some of these are cheap to do; others are less so.
+Fortunately the HF seems willing to help.
+
+*But whatever we do here will be a step forward* from our current, unsatisfactory situation.
+Moreover, they will help with CI for changes to GHC itself! (It is rather *more* likely that a commit to GHC's simplifier will cause a perf regression in some package, than a commit to ``ghc-internals``.)
+
+Discussion
+==========
+
+GHC Proposals process
+---------------------
+
+Some GHC proposals (a minority) directly affect the existing API of ``base``, and are not simply additions that can be exposed in ``ghc-experimental``.
+It is unproductive for the GHC Steering Committee to have a long discussion, accept the proposals, and only *then* involve the CLC.
+
+We propose that:
+
+- A GHC Proposal should advertise, in a separate section:
+
+  -  What changes, if any, it make to ``ghc-experimental``
+
+  -  What changes, if any, it make to ``base``
+
+- If there are any such changes, the author (and shepherd) should explicitly invite the CLC to participate in the discussion about the proposal.
+  The CLC will devote some effort to participating and, in the case of changes to ``base``, will subsequently hold a non-binding vote.
+
+- Approval of the proposal (by the GHC Steering Committee, with the non-binding vote of CLC) is not a guarantee that the final implementation will land;
+  that depends on the implementation being well engineered etc (GHC team);
+  and the implementor should make an explicit proposal to the CLC specifying the precise changes.
+
+Abstraction leakage
+-------------------
+
+We may foresee a couple of ways in which changes in ``ghc-internals`` could become client visible:
+
+- Occasionally, an error message may mention a fully qualified name for an out-of-scope identifier.
+  For example (GHC test ``mod153``)::
+
+    Ambiguous occurrence ‘id’
+               It could refer to either ‘Prelude.id’,
+                            imported from ‘Prelude’ at mod153.hs:2:8
+                            (and originally defined in ‘GHC.Base’)
+                         or ‘M.id’, defined at mod153.hs:2:21
+
+  The "originally defined in" mentions a module; and if that module is in a package that is not imported, GHC will package-qualify the module name.
+  And seeing ``ghc-internals:GHC.Base`` is perhaps less nice.
+  This is not a new problem: we already package-qualify modules in ``ghc-prim``.
+  One solution is to remove the "originally defined in.." parenthesis for types and functions that would require such package qualification.
```

Comment:
IMO a lot of this would be helped by tracking the import/re-export provenance more directly. "Imported from A and originally defined in E" is nice but it's also often very nice to know the exact chain of imports and re-exports that brought something into scope. GHC could then have a flag to trim this provenance information to the direct dependencies of a package. That way, if you find yourself depending on `ghc-internals:SecretType`, you can see exactly how it got introduced into your codebase.

Reply:
In asking https://github.com/ghc-proposals/ghc-proposals/pull/595#issuecomment-1607844684 I was wondering something similar --- ideally we we do now know the entire provenance of imports and (re)exports, and there are a number of things we can do with it.